expression/builtin_string.go

```diff
@@ -3976,3 +3976,146 @@ func (b *builtinWeightStringSig) evalString(row chunk.Row) (string, bool, error)
 	}
 	return string(ctor.Key(str)), false, nil
 }
+
+type translateFunctionClass struct {
+	baseFunctionClass
+}
+
+// getFunction sets translate built-in function signature.
+// The syntax of translate in Oracle is 'TRANSLATE(expr, from_string, to_string)'.
+func (c *translateFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETString, types.ETString, types.ETString, types.ETString)
+	if err != nil {
+		return nil, err
+	}
+	argType := args[0].GetType()
+	bf.tp.Flen = argType.Flen
+	SetBinFlagOrBinStr(argType, bf.tp)
+	if types.IsBinaryStr(args[0].GetType()) || types.IsBinaryStr(args[1].GetType()) || types.IsBinaryStr(args[2].GetType()) {
+		sig := &builtinTranslateBinarySig{bf}
+		return sig, nil
+	}
+	sig := &builtinTranslateUTF8Sig{bf}
+	return sig, nil
+}
+
+type builtinTranslateBinarySig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinTranslateBinarySig) Clone() builtinFunc {
+	newSig := &builtinTranslateBinarySig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+// evalString evals a builtinTranslateSig, corresponding to translate(srcStr, fromStr, toStr)
+// See https://docs.oracle.com/cd/B19306_01/server.102/b14200/functions196.htm
+func (b *builtinTranslateBinarySig) evalString(row chunk.Row) (d string, isNull bool, err error) {
+	var (
+		srcStr, fromStr, toStr     string
+		isFromStrNull, isToStrNull bool
+		tgt                        []byte
+	)
+	srcStr, isNull, err = b.args[0].EvalString(b.ctx, row)
+	if isNull || err != nil {
+		return d, isNull, err
+	}
+	fromStr, isFromStrNull, err = b.args[1].EvalString(b.ctx, row)
+	if isFromStrNull || err != nil {
+		return d, isFromStrNull, err
+	}
+	toStr, isToStrNull, err = b.args[2].EvalString(b.ctx, row)
+	if isToStrNull || err != nil {
+		return d, isToStrNull, err
+	}
+	mp := buildTranslateMap4Binary([]byte(fromStr), []byte(toStr))
+	for _, charSrc := range []byte(srcStr) {
+		if charTo, ok := mp[charSrc]; ok {
+			if charTo != math.MaxUint8 {
+				tgt = append(tgt, charTo)
```

Comment:
Looks like not correct.

    mysql> select translate(x'ff', x'ff', x'ff');
    +----------------------------------------------------------------+
    | translate(x'ff', x'ff', x'ff')                                 |
    +----------------------------------------------------------------+
    | 0x                                                             |
    +----------------------------------------------------------------+
    1 row in set (0.00 sec)